packages/context/src/provider.ts

```diff
@@ -0,0 +1,119 @@
+// Copyright IBM Corp. 2013,2017. All Rights Reserved.
+// Node module: loopback
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Context} from './context';
+import {BoundValue, ValueOrPromise} from './binding';
+import {Constructor, instantiateClass} from './resolver';
+
+/**
+ * Provider of a value
+ */
+export interface Provider<T> {
+  value(ctx?: Context): ValueOrPromise<T>;
```

Comment:
Oh, I found an earlier @raymondfeng's comment saying "the optional ctx can be injected too". If that's the case, then we need to ensure all Provider specializations support invocation of `value()` with not ctx argument, and include tests to verify these use cases.

TBH, I think this is not worth the effort, because `Provider.value()` is called by our context framework and the `ctx` argument is always set.